### PR TITLE
8075806: divideExact is missing in java.lang.Math

### PR DESCRIPTION
**Pull request by bplb**
Please consider this proposal to add `divideExact()` methods for integral data types to `java.lang.Math` thereby rounding out "exact" support to all four basic arithmetic operations.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8075806](https://bugs.openjdk.java.net/browse/JDK-8075806): divideExact is missing in java.lang.Math

### Reviewers
 * [Joe Darcy](https://openjdk.java.net/census#darcy) (@jddarcy - **Reviewer**) ⚠️ Review applies to 5093a2e877adcaff76b0aa65d5a9b290170f2882

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4770/head:pull/4770` \
`$ git checkout pull/4770`

Update a local copy of the PR: \
`$ git checkout pull/4770` \
`$ git pull https://git.openjdk.java.net/jdk pull/4770/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4770`

View PR using the GUI difftool: \
`$ git pr show -t 4770`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4770.diff">https://git.openjdk.java.net/jdk/pull/4770.diff</a>

</details>

### CODE REVIEW

**Comment by bplb** [comment]
Note: The `StrictMath` equivalents were intentionally omitted until the review converges to a consensus.

File: src/java.base/share/classes/java/lang/Math.java [hunk 1]
```diff
@@ -1007,6 +1002,53 @@ public static long multiplyExact(long x, long y) {
         return r;
     }
 
+    /**
+     * Returns the quotient of the arguments, throwing an exception if the
+     * result overflows an {@code int}.  Such overflow can occur if and only
+     * if either {@code y} is zero, or both {@code x} is
```
Comment (jddarcy) [inline src/java.base/share/classes/java/lang/Math.java:1008]: I think the divide by zero case should be discussed separately from overflow.

**Comment by jddarcy** [comment]
Are there examples in the JDK or its tests where a method with this definition would be used?

**Comment by bplb** [comment]
Thanks. I was thinking of leveraging the `negateExact()` intrinsic instead:

```
        // Leverage a potential intrinsic of negateExact()
        return y == -1 ? Math.negateExact(x) : x/y;
```

**Comment by bplb** [comment]
@rgiulietti Actually your trick is the fastest: `yours > via negateExact() > original`. Thanks!

File: src/java.base/share/classes/java/lang/Math.java (hunk 1 above)
Comment (bplb) [reply src/java.base/share/classes/java/lang/Math.java]: Fixed.

**Comment by bplb** [comment]
> Are there examples in the JDK or its tests where a method with this definition would be used?

I have not identified any as yet. I did verify however that there are no uses in `open/src/**/*.java` of either `absExact()` or `negateExact()`, only their implementations in `[Strict]Math`.

File: src/java.base/share/classes/java/lang/Math.java [hunk 2]
```diff
@@ -1007,6 +1002,58 @@ public static long multiplyExact(long x, long y) {
...
 
+    /**
+     * Returns the quotient of the arguments, throwing an exception if the
+     * result overflows an {@code int}.  Such overflow occurs in this method if
+     * {@code x} is {@link Integer#MIN_VALUE} and {@code y} is {@code -1}.
+     * In contrast, if {@code Integer.MIN_VALUE / -1} were evaluated directly,
+     * the result would be {@code Integer.MIN_VALUE} and no exception would be
+     * thrown.
+     * <p>
+     * If {@code y} is zero, overflow also occurs as in direct evaluation and
```
Comment (jddarcy) [inline src/java.base/share/classes/java/lang/Math.java:1013]: To clarify my earlier comments, I don't think the divide-by-zero case should be described as "overflow". An overflow condition generally could be handled if the result format had greater range. That is not true of dividing by an exact zero.

However, the methods should still throw an exception for divide by zero.

Comment (rgiulietti) [reply src/java.base/share/classes/java/lang/Math.java:1013]: @bplb More specifically, a int/long division by 0 throws ArithmeticException("/ by zero")

File: test/jdk/java/lang/Math/ExactArithTests.java [hunk 3]
```diff
@@ -132,6 +132,39 @@ static void testIntegerExact(int x, int y) {
             }
         }
 
+        boolean exceptionExpected = false;
```
Comment (jddarcy) [inline test/jdk/java/lang/Math/ExactArithTests.java:135]: Please add "divideExact" to the list of methods tested by 
static void testIntegerExact(int x, int y) 
and similar change for long.

**Review by jddarcy — APPROVED** [review]

**Pull request closed without merging**